src/geo.c

```
@@ -635,22 +672,26 @@ void georadiusGeneric(client *c, int srcKeyIndex, int flags) {
     if (storekey && (withdist || withhash || withcoords)) {
         addReplyErrorFormat(c, "%s is not compatible with WITHDIST, WITHHASH and WITHCOORD options",
                             flags & GEOSEARCHSTORE ? "GEOSEARCHSTORE" : "STORE option in GEORADIUS");
+        geoPolygonPointsFree(&shape);
         return;
     }
 
-    if ((flags & GEOSEARCH) && !(frommember || fromloc)) {
+    if ((flags & GEOSEARCH) && !(frommember || fromloc) && !bypolygon) {
         addReplyErrorFormat(c, "exactly one of FROMMEMBER or FROMLONLAT can be specified for %s",
```

What error message do we get for `(frommember || fromloc) && bypolygon`? I'd expect something like "FROMMEMBER and FROMLONLAT can't be used with BYPOLYGON".

Please add a test case to cover this too.

---

Or should we accept an optional FROM for polygons? We could use it for WITHDIST instead of the centroid if it's provided... WDYT?

---

> What error message do we get for (frommember || fromloc) && bypolygon? I'd expect something like "FROMMEMBER and FROMLONLAT can't be used with BYPOLYGON".
Please add a test case to cover this too.

Currently, `-ERR syntax error` gets returned to the client because of this catch all block here in `void georadiusGeneric(`:
```
            } else {
                addReplyErrorObject(c, shared.syntaxerr);
                geoPolygonPointsFree(&shape);
                return;
            }
```

I haver a test case for this already in geo.tcl over here:
```
        # Error Case - Users cannot use BYPOLYGON along with FROMLONLAT or FROMMEMBER.
        catch {r GEOSEARCH points BYPOLYGON 3 151.2039101016274 -33.87446472104431 151.2132606898065 -33.882942295615855 151.2132606898065 -33.882942295615855 FROMLONLAT 151.2229523062706 -33.883973760201364} e
        assert_match {*ERR syntax error*} $e
        catch {r GEOSEARCH points BYPOLYGON 3 151.2039101016274 -33.87446472104431 151.2132606898065 -33.882942295615855 151.2132606898065 -33.882942295615855 FROMMEMBER memberA} e
        assert_match {*ERR syntax error*} $e
```

---

> Or should we accept an optional FROM for polygons? We could use it for WITHDIST instead of the centroid if it's provided... WDYT?

IMO, using `FROMLONLAT` or `FROMMEMBER` to calculate the distance might be slightly confusing since they are used as the starting point for BYRADIUS and BYBOX.

In the case of BYPOLYGON, the withdist currently uses the centroid which was not directly provided by the user, but is calculated as a result of polygon vertices. 

Since the search results are all based on the centroid, using different distance calculation (starting from a different member) might be confusing I feel.


If we want to use an external member for the distance calculation, we can probably consider adding a new `DISTFROM` parameter that can apply to all 3 searches - BOX, RADIUS, BYPOLYGON

---

Use the centroi. 👍 

Regarding the syntax error, i guess a specific error can be nicer, since the syntax in the json/doc/hints now appears to allow FROM with BYPOLYGON. I don't feel strongly though.